### ISSUE TITLE
Make `home.sessionVariables` work with nushell

### ISSUE THREAD

**Issue by Kreyren** [nix-community/home-manager#4313]
### Description

Currently i have to set the following:

```nix
  home.sessionVariables = {
    SHELL = "nu";
    TERM = "alacritty";
    EDITOR = "vim";
  };

  programs.nushell.extraEnv = ''
    # FIXME(Krey): This should be handled in `home.sessionVariables`
    $env.SHELL = "nu";
    $env.TERM = "alacritty"
    $env.EDITOR = "vim";

    # Include local binaries
    $env.PATH = ($env.PATH | split row (char esep) | append ($env.HOME | path join '.local/bin'));
  ''; 
```

Because `home.sessionVariables` doesn't export the variables for `nushell` -> Make it work with nushell too?

Expected workflow:

```diff
  home.sessionVariables = {
    SHELL = "nu";
    TERM = "alacritty";
    EDITOR = "vim";
  };

  programs.nushell.extraEnv = ''
-    # FIXME(Krey): This should be handled in `home.sessionVariables`
-    $env.SHELL = "nu";
-    $env.TERM = "alacritty"
-    $env.EDITOR = "vim";
-
    # Include local binaries
    $env.PATH = ($env.PATH | split row (char esep) | append ($env.HOME | path join '.local/bin'));
  ''; 
```

**Comment by Suyashtnt** [comment]
It's been a few months now. Are there any plans for this?

**Comment by bluskript** [comment]
quick fix that I haven't had time to cook into a PR / probably has compat problems:
```nix
    environmentVariables = builtins.mapAttrs (name: value: "\"${builtins.toString value}\"") config.home.sessionVariables;
```

**Comment by heisfer** [comment]
Maybe we could make use of https://github.com/tesujimath/nu_plugin_bash_env this plugin?

**Comment by fehnomenal** [comment]
This is still relevant.

After https://github.com/NixOS/nixpkgs/pull/343036 is merged, I can create a PR because I have this setup working.

**Comment by hey2022** [comment]
> quick fix that I haven't had time to cook into a PR / probably has compat problems:
> 
> ```nix
>     environmentVariables = builtins.mapAttrs (name: value: "\"${builtins.toString value}\"") config.home.sessionVariables;
> ```

Buidling upon @bluskript quick fix, this changes references to environment variables such as $PATH to $env.PATH

```nix
environmentVariables = builtins.mapAttrs (
  name: value: "\"${builtins.replaceStrings [ "$" ] [ "$env." ] (builtins.toString value)}\""
) config.home.sessionVariables;
```

**Comment by jaredmontoya** [comment]
@fehnomenal `bash-env-json` 0.9.1 is merged into nixpkgs at the moment of writing, track here: https://nixpk.gs/pr-tracker.html?pr=358140

0.9.1 fails on files that have someting liike `PATH=/some/path/that/has/no/jq/sed/or/coreutils` but it is enough to start implementing home-manager session vars.

0.9.2 with a fix for this will be merged soon: https://nixpk.gs/pr-tracker.html?pr=358950

Here is how I am using it currently:
`bash-env-nu.nix`:
```nix
{ lib, bash-env-json }:

# nu
''
  module bash-env {
    export def main [
      path?: string
      --export: list
      --shellvars (-s)
      --fn (-f): list
    ] {
      let fn_args = if ($fn | is-not-empty) {
        ['--shellfns' ($fn | str join ',')]
      } else {
        []
      }

      let path_args = if $path != null {
        [($path | path expand)]
      } else {
        []
      }

      let raw = ($in | str join "\n") | ${lib.getExe bash-env-json} ...($fn_args ++ $path_args) | complete
      let raw_json = $raw.stdout | from json

      let error = $raw_json | get -i error
      if $error != null {
        error make { msg: $error }
      } else if $raw.exit_code != 0 {
        error make { msg: $"unexpected failure from bash-env-json ($raw.stderr)" }
      }

      if ($export | is-not-empty) {
        print "warning: --export is deprecated, use --shellvars(-s) instead"
        let exported_shellvars = ($raw_json.shellvars | select -i ...$export)
        $raw_json.env | merge ($exported_shellvars)
      } else if $shellvars or ($fn | is-not-empty) {
        $raw_json
      } else {
        $raw_json.env
      }
    }
  }
  use bash-env
''
```

Here is an example of home-manager session vars + dircolors integration:
```nix
      nushell = {
        enable = true;

        envFile.text = # nu
          ''
            source ${pkgs.writeText "bash-env-nu" (pkgs.callPackage ./bash-env-nu.nix { })}

            bash-env ${config.home.profileDirectory}/etc/profile.d/hm-session-vars.sh | load-env
            dircolors | bash-env | load-env
          '';
    };

**Comment by jaredmontoya** [comment]
Update: I am currently working on a PR for home-manager that should improve nushell support.
https://github.com/jaredmontoya/home-manager/tree/nushell-bash-env
And I need help.

In the previous comment I showed how I implemented run-time bash environment translation in my nushell configuration.
Unfortunately this drastically increases nushell startup times.

Fortunately the solution is to simply **do the translation at build-time**.
Here is how to do the exact same thing as in the previous comment but at build-time:

(`pkgs.bash-env-nu` is added through an overlay to avoid multiple calls:  `pkgs.callPackage ./bash-env-nu.nix { }`)
```nix
let
  hmSessionVarsNu = pkgs.runCommand "hm-session-vars.nu" { } ''
    echo "if ('__HM_SESS_VARS_SOURCED' not-in \$env) {''$(${lib.getExe config.programs.nushell.package} -c "
      source ${pkgs.bash-env-nu}
      bash-env ${config.home.sessionVariablesPackage}/etc/profile.d/hm-session-vars.sh | to nuon
    ") | load-env}" >> "$out"
  '';

  dircolorsIntegration = pkgs.runCommand "dircolors.nu" { } ''
    echo "''$(${lib.getExe config.programs.nushell.package} -c "
      source ${pkgs.bash-env-nu}
      ${lib.getExe' pkgs.coreutils "dircolors"} | bash-env | to nuon
    ") | load-env" >> "$out"
  '';
in
```

```nix
      nushell = {
        enable = true;

        envFile.text = # nu
          ''
            source ${hmSessionVarsNu}
            source ${dircolorsIntegration}
          '';
    };
```

As you can see in https://github.com/jaredmontoya/home-manager/commit/b62095d8fa7b63a6836ead1c53980728e9732441
I already implemented home-manager session variables build-time translation and also improved dircolors module.

But there is a problem, I just call `bash-env-nu` package that I added to the nushell home-manager module directory which is not ideal if I need to access it in multiple places.

`bash-env-nu` is just a small nu script that defines a nu `bash-env` module that wraps around `bash-env-json` that I added to nixpkgs already, so I think `bash-env-nu` belongs somewhere in home-manager rather than nixpkgs.

In my config I just use my overlay system to add this package, but I don't know what is the best way to add it to home-manager.

In order to finish this PR I will need suggestions of people more experienced with the structure of home-manager.

**Comment by stepbrobd** [comment]
Quick question, can we use the function described [here](https://www.nushell.sh/cookbook/foreign_shell_scripts.html#capturing-the-environment-from-a-foreign-shell-script) than using a plugin?

I tested the code snippet and it worked quite well, except for the fact that, as I quote:

> A caveat for this approach is that it requires all changed environment variables not to include newline characters, as the UNIX env output is not cleanly parseable in that case.

If the plugin clearly addresses the caveat mentioned above then nvm

**Comment by jaredmontoya** [comment]
@stepbrobd it's not really a plugin, there is a difference between a nushell PLUGIN and a nushell MODULE.
You can learn more by executing `use --help` in nushell.

previously [bash-env-nushell](https://github.com/tesujimath/bash-env-nushell) was a plugin but now @tesujimath encourages to use a module implementation of `bash-env` because of how much more simpler and compatible with different nushell versions it is, the file is called `bash-env.nu` in the root of the repository. Rust code is just plugin source code that is kept there for backwards compatibility.

Unfortunately I can't answer your question on whether bash-env-json resolves this caveat, but I think @tesujimath can.

I am not sure if I understood correctly what the nushell documentation meant but a quick test like
```nushell
"export tEsT='multi
line'"| bash-env-json
```
returns `jq: parse error: Expected separator between values at line 2, column 6`

**Comment by jaredmontoya** [comment]
@stepbrobd I am happy to announce that I just solved this problem once and for all
(as long as your environment variables do not contain NUL bytes for some reason)

Here is my modified version of the nushell capture-foreign-env function:
```nushell
# Returns a record of changed env variables after running a non-nushell script's contents (passed via stdin), e.g. a bash script you want to "source"
def capture-foreign-env [
  --shell (-s): string = /bin/sh
  # The shell to run the script in
  # (has to support '-c' argument and POSIX 'env', 'echo', 'eval' commands)
  --arguments (-a): list<string> = []
  # Additional command line arguments to pass to the foreign shell
] {
  let script_contents = $in;
  let env_out = with-env { SCRIPT_TO_SOURCE: $script_contents } {
    ^$shell ...$arguments -c `
    env -0
    echo -n '<ENV_CAPTURE_EVAL_FENCE>'
    eval "$SCRIPT_TO_SOURCE"
    echo -n '<ENV_CAPTURE_EVAL_FENCE>'
    env -0 -u _ -u _AST_FEATURES -u SHLVL` # Filter out known changing variables
  }
  | split row '<ENV_CAPTURE_EVAL_FENCE>'
  | {
    before: ($in | first | str trim --char (char -i 0) | split row (char -i 0))
    after: ($in | last | str trim --char (char -i 0) | split row (char -i 0))
  }

  $env_out.after
  | where { |line| $line not-in $env_out.before } # Only get changed lines
  | parse "{key}={value}"
  | transpose --header-row --as-record
}
```

And here is how it DOES NOT fail on environment variables with newlines:
```nu
"export tEsT='mult
∙ iline'"| capture-foreign-env
╭──────┬───────╮
│ tEsT │ mult  │
│      │ iline │
╰──────┴───────╯
```

**Comment by stepbrobd** [comment]
> @stepbrobd I am happy to announce that I just solved this problem once and for all (as long as your environment variables do not contain NUL bytes for some reason)

Awesome work!

Now it seems that we only need to piece some options together, and pull in environment variables from from `home.sessionVariables` and `osConfig` (since there's no equivalent of `/etc/bashrc`, `/etc/zshrc`, `/etc/zprofile`, etc. in Nushell).

With Nushell, settings from `/etc/...` are not automatically sourced like how other shells automatically inherit from system-level settings, then user-level. For example:

- nixpkgs:
  - https://github.com/NixOS/nixpkgs/blob/8bfc09287d96a38707d7a7c12a2e55f8a172d153/nixos/modules/programs/bash/bash.nix#L123
  - https://github.com/NixOS/nixpkgs/blob/8bfc09287d96a38707d7a7c12a2e55f8a172d153/nixos/modules/programs/bash/bash.nix#L130
  - https://github.com/NixOS/nixpkgs/blob/f56ca361905c5c47c51c4b9556c33be31dbfb360/nixos/modules/programs/zsh/zsh.nix#L185-L187
  - https://github.com/NixOS/nixpkgs/blob/8bfc09287d96a38707d7a7c12a2e55f8a172d153/nixos/modules/programs/xonsh.nix#L45-L58

- nix-darwin:
  - https://github.com/LnL7/nix-darwin/blob/71a3a075e3229a7518d76636bb762aef2bcb73ac/modules/programs/zsh/default.nix#L139-L142
  - https://github.com/LnL7/nix-darwin/blob/71a3a075e3229a7518d76636bb762aef2bcb73ac/modules/programs/zsh/default.nix#L167
  - https://github.com/LnL7/nix-darwin/blob/71a3a075e3229a7518d76636bb762aef2bcb73ac/modules/programs/zsh/default.nix#L194

**Comment by jaredmontoya** [comment]
@stepbrobd I find it very weird but even though I use `nushell` as my user login shell, I have `$env.__NIXOS_SET_ENVIRONMENT_DONE` set to 1 (which means it somehow inherited NixOS environment from something.)

do you know why that could be the case?

I use `tuigreet` to start my session
```nix
greetd = {
      enable = true;
      settings = {
        default_session = {
          command = builtins.concatStringsSep " " [
            (lib.getExe pkgs.greetd.tuigreet)
            "--theme 'border=blue;text=cyan;prompt=green;time=red;action=blue;button=yellow;container=black;input=red'"
            "--remember"
            "--user-menu"
            "--time"
            "--asterisks"
            "--cmd 'Hyprland'"
          ];
          user = "username";
        };
      };
    };
```

the "username" user has `shell = pkgs.nushell;`

**Comment by stepbrobd** [comment]
> @stepbrobd I find it very weird but even though I use `nushell` as my user login shell, I have `$env.__NIXOS_SET_ENVIRONMENT_DONE` set to 1 (which means it somehow inherited NixOS environment from something.)

That's what I have with NixOS as well (I use gtkgreet through greetd), but on my nix-darwin machine setting `users.users.<name>.shell = pkgs.nushell` won't cut it. It just starts from a blank Nushell with absolutely zero config and all the paths and envvars are not set. I have to run `exec zsh -c nu` to get all the configs back, which I think is absolutely weird :/

I did some digging on the implementation of `users.users.<name>.shell` in nixpkgs and nix-darwin, but I don't think they do anything special relating to setting all the paths and envvars

> do you know why that could be the case?

Absolutely zero clue 😂

**Comment by iniw** [comment]
Want to add that shell aliases should work too. Currently I do this:
```nix
nushell = {                                                        
  # ...
  shellAliases = config.home.shellAliases;                         
  environmentVariables = config.home.sessionVariables;             
};                                                                 
```
Which is fine but should really be integrated directly into home-manager.

**Comment by jaredmontoya** [comment]
> Want to add that shell aliases should work too. Currently I do this:
> 
> nushell = {                                                        
>   shellAliases = config.home.shellAliases;                         
>   environmentVariables = config.home.sessionVariables;             
> };                                                                 
> 
> Which is fine but should really be integrated directly into home-manager.

I agree about shellAliases: https://github.com/nix-community/home-manager/pull/4300#issuecomment-2500132666
documentation says that only aliases comptible with ALL shells must be set to `home.shellAliases`, otherwise they should go in `programs.bash.shellAliases` if bash specific syntax is used. And ALL shells doesn't mean bash, zsh, fish

**Comment by stepbrobd** [comment]
Still important

**Comment by mountgrabber** [comment]
Still important

**Comment by kaptcha0** [comment]
Still having this issue to this day. For whatever reason, it was working briefly (I can't remember when) but no longer.

**Comment by dabljues** [comment]
Still relevant. The trick with `environmentVariables = config.home.sessionVariables;` works only partially:

Home manager can generate something like this:

```sh
export XDG_DATA_DIRS="${NIX_STATE_DIR:-/nix/var/nix}/profiles/default/share:/home/dabljues/.nix-profile/share:/usr/share/ubuntu:/usr/local/share:/usr/share:/var/lib/snapd/desktop${XDG_DATA_DIRS:+:$XDG_DATA_DIRS}"
```

If I do it in bash, this would expand into:

```sh
XDG_DATA_DIRS=/nix/var/nix/profiles/default/share:/home/dabljues/.nix-profile/share:/usr/share/ubuntu:/usr/local/share:/usr/share:/var/lib/snapd/desktop
```
However, in nushell:

```sh
➜ $env.XDG_DATA_DIRS
${NIX_STATE_DIR:-/nix/var/nix}/profiles/default/share:/home/dabljues/.nix-profile/share:/usr/share/ubuntu:/usr/local/share:/usr/share:/var/lib/snapd/desktop${XDG_DATA_DIRS:+:$XDG_DATA_DIRS}
```

This is obviously not right. When I launch vim I get spammed with: `E79: Cannot expand wildcards`
